kvbc/include/kvbc_adapter/kv_blockchain_adapter.hpp

```diff
@@ -0,0 +1,158 @@
+
+namespace concord::kvbc::adapter {
+class KeyValueBlockchain {
+ public:
+  template <typename OPTIONS,
+            typename = std::enable_if_t<std::is_same_v<OPTIONS, NativeClient::DefaultOptions> ||
+                                        std::is_same_v<OPTIONS, NativeClient::ExistingOptions> ||
+                                        std::is_same_v<OPTIONS, NativeClient::UserOptions>>>
+  [[maybe_unused]] explicit KeyValueBlockchain(const std::string& db_path,
+                                               bool read_only,
+                                               bool link_st_chain,
+                                               const OPTIONS& options) {
+    // The logic of choosing type of block chain will happen here.
+    if (bftEngine::ReplicaConfig::instance().kvBlockchainVersion == BLOCKCHAIN_VERSION::CATEGORIZED_BLOCKCHAIN) {
+      auto kv = std::make_shared<CatBCimpl::element_type>(NativeClient::newClient(db_path, read_only, options),
+                                                          link_st_chain);
+      kvbc_.emplace<CatBCimpl>(kv);
+    }
+  }
+
+  // Creates a key-value blockchain.
+  // If `category_types` is nullopt, the persisted categories in storage will be used.
+  // If `category_types` has a value, it should contain all persisted categories in storage at a minimum. New ones
+  // will be created and persisted. Users are required to pass a value for `category_types` on first construction
+  // (i.e. a new blockchain) in order to specify the categories in use. Failure to do so will generate an exception.
+  explicit KeyValueBlockchain(
+      const std::shared_ptr<concord::storage::rocksdb::NativeClient>& native_client,
+      bool link_st_chain,
+      const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>>& category_types = std::nullopt) {
+    if (bftEngine::ReplicaConfig::instance().kvBlockchainVersion == BLOCKCHAIN_VERSION::CATEGORIZED_BLOCKCHAIN) {
+      auto kv = std::make_shared<CatBCimpl::element_type>(native_client, link_st_chain, category_types);
+      kvbc_.emplace<CatBCimpl>(kv);
+    }
+  }
+
+  // The noop converter returns the input string as is, without modifying it.
+  static const categorization::Converter kNoopConverter;
+
+  /////////////////////// Add Block ///////////////////////
+
+  BlockId addBlock(categorization::Updates&& updates);
+
+  /////////////////////// Delete block ///////////////////////
+
+  bool deleteBlock(const BlockId& blockId);
+  void deleteLastReachableBlock();
+
+  /////////////////////// Raw Blocks ///////////////////////
+
+  // Adds raw block and tries to link the state transfer blockchain to the main blockchain
+  // If linkSTChainFrom is true, try to remove blocks form the state transfer chain to the blockchain
+  void addRawBlock(const categorization::RawBlock& block, const BlockId& block_id, bool lastBlock = true);
+  std::optional<categorization::RawBlock> getRawBlock(const BlockId& block_id) const;
```

Comment:
Those methods are not part of an interface, do we want to add non-interface methods to the adapter? 

- Who will call them?
- How they'll be supported by the V4 implementation? they are specific to the categorized storage

Reply:
Thanks, @cloudnoize, I am changing this all as discussed and making sure that the Adapter implements the correct interface.